pkg/koordlet/runtimehooks/nri/server.go

```diff
@@ -0,0 +1,189 @@
+package nri
+
+import (
+	"context"
+	"fmt"
+	"github.com/containerd/nri/pkg/api"
+	"github.com/containerd/nri/pkg/stub"
+	"github.com/koordinator-sh/koordinator/pkg/koordlet/resourceexecutor"
+	"github.com/koordinator-sh/koordinator/pkg/koordlet/runtimehooks/hooks"
+	"github.com/koordinator-sh/koordinator/pkg/koordlet/runtimehooks/protocol"
+	rmconfig "github.com/koordinator-sh/koordinator/pkg/runtimeproxy/config"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/yaml"
+	"strings"
+)
+
+type nriconfig struct {
+	LogFile       string   `json:"logFile"`
+	Events        []string `json:"events"`
+	AddAnnotation string   `json:"addAnnotation"`
+	SetAnnotation string   `json:"setAnnotation"`
+	AddEnv        string   `json:"addEnv"`
+	SetEnv        string   `json:"setEnv"`
+}
+
+type Options struct {
+	// support stop running other hooks once someone failed
+	PluginFailurePolicy rmconfig.FailurePolicyType
+	DisableStages       map[string]struct{}
+	Executor            resourceexecutor.ResourceUpdateExecutor
+}
+
+type NriServer struct {
+	stub            stub.Stub
+	mask            stub.EventMask
+	options         Options // server options
+	runPodSandbox   func(*NriServer, *api.PodSandbox, *api.Container) error
+	createContainer func(*NriServer, *api.PodSandbox, *api.Container) (*api.ContainerAdjustment, []*api.ContainerUpdate, error)
+	updateContainer func(*NriServer, *api.PodSandbox, *api.Container) ([]*api.ContainerUpdate, error)
+}
+
+var (
+	_ = stub.ConfigureInterface(&NriServer{})
+	_ = stub.SynchronizeInterface(&NriServer{})
+	_ = stub.RunPodInterface(&NriServer{})
+	_ = stub.CreateContainerInterface(&NriServer{})
+	_ = stub.UpdateContainerInterface(&NriServer{})
+
+	pluginName = "koordlet_nri"
+	pluginIdx  = "00"
+	events     = "RunPodSandbox,CreateContainer,UpdateContainer"
+	cfg        nriconfig
+	opts       []stub.Option
+	err        error
+)
+
+func NewNriServer(opt Options) (*NriServer, error) {
+	if Enabled() {
+		opts = append(opts, stub.WithPluginName(pluginName))
+		opts = append(opts, stub.WithPluginIdx(pluginIdx))
+		p := &NriServer{options: opt}
+		if p.mask, err = api.ParseEventMask(events); err != nil {
+			klog.Errorf("failed to parse events: %v", err)
+		}
+		cfg.Events = strings.Split(events, ",")
+
+		if p.stub, err = stub.New(p, append(opts, stub.WithOnClose(p.onClose))...); err != nil {
+			klog.Errorf("failed to create plugin stub: %v", err)
+		}
+
+		return p, err
+	}
+	return nil, nil
+}
+
+func (s *NriServer) Start() error {
+	go func() {
+		klog.Info("-------nri start-----------")
+		err := s.stub.Run(context.Background())
```

Comment:
from https://github.com/containerd/nri/blob/main/pkg/stub/stub.go, it is possible that stub.New return nil. so it may be better to check 
if s.stub != nil {
    err := s.stub.Run(context.Background())
    ...
}